tests/mock_crust/churn.rs

```diff
@@ -90,24 +85,61 @@ fn add_nodes_and_poll<R: Rng>(
             pfx
         })
         .collect();
+
+    let mut added_nodes = Vec::new();
     while !prefixes.is_empty() {
+        let proxy_index = if nodes.len() > nodes[0].chain().min_sec_size() {
+            gen_range(rng, 0, nodes.len())
+        } else {
+            0
+        };
+        let bootstrap_config =
+            BootstrapConfig::with_contacts(&[nodes[proxy_index].handle.endpoint()]);
         let node = TestNode::builder(&network)
             .bootstrap_config(bootstrap_config.clone())
             .create();
         if let Some(&pfx) = prefixes.iter().find(|pfx| pfx.matches(&node.name())) {
             assert!(prefixes.remove(&pfx));
-            if allow_add_failure && !rng.gen_weighted_bool(prefixes.len() as u32) {
+            if skip_some_prefixes && !rng.gen_weighted_bool(prefixes.len() as u32) {
                 continue;
             }
-            nodes.push(node);
+            added_nodes.push(node);
         }
     }
 
-    poll_and_resend(&mut nodes, &mut []);
+    added_nodes.drain(..).for_each(|added_node| {
+        let index = gen_range(rng, 1, nodes.len() + 1);
+        nodes.insert(index, added_node);
+    });
+
+    nodes
+        .iter()
+        .enumerate()
+        .filter_map(|(index, node)| {
+            if !node.inner.is_node() {
+                Some(index)
+            } else {
+                None
+            }
+        })
+        .collect::<BTreeSet<_>>()
+}
 
+/// Checks if the given indices have been accepted to the network.
+/// allow_add_failure: Allows nodes to fail getting accepted.
+fn check_added_indices<R: Rng>(
```

Comment:
The name of this function is misleading: it does much more than what the name says. It:
1. Checks all new nodes were successfully added
2. Removes failed nodes
3. Clears relocation overrides
4. Polls the nodes
5. Shuffles the nodes

So either it should be given more suitable name or maybe better be split into several functions each with clear responsibility.